### PR TITLE
Renew csr on migration

### PR DESCRIPTION
**Pull request by PietroPasotti**
## Issue
when upgrading and migrating from peer-backed vault to secrets-backed vault, certhandler regenerates the privkey because it can't find one in the new (current) vault, but it migrates the old csr over.
Consequently the csr is signed with an outdated and lost privkey. 

## Solution
on upgrade, after migrating the vault, if we have a csr but NO private key in the vault, we know the CSR is signed by an outdated privkey. Renew it.

[Tandem pr](https://github.com/canonical/traefik-k8s-operator/pull/349)

## Testing Instructions
```
juju add-model mytestmodel
juju deploy cos-lite --trust --channel=stable
# refresh traefik from [the tandem pr branch]
curl -v -k https://<TRAEFIK IP>/mytestmodel-prometheus-0/api/v1/targets
```

### CODE REVIEW

**Comment by PietroPasotti** [comment]
Tandem PR CI is green (modulo forward_auth tests).

File: lib/charms/observability_libs/v1/cert_handler.py [hunk 1]
```diff
@@ -349,21 +349,36 @@ def __init__(
...
+            self._generate_csr(renew=True)
+
     def _migrate_vault(self):
         peer_backend = _RelationVaultBackend(self.charm, relation_name="peers")
 
         if self._check_juju_supports_secrets():
             # we are on recent juju
             if self.vault.retrieve():
                 # we already were on recent juju: nothing to migrate
+                logger.debug("vault is already populated. Nothing to migrate...")
```
Comment (sed-i) [inline lib/charms/observability_libs/v1/cert_handler.py:371]:
```suggestion
                logger.debug("Private key is already stored as a juju secret. Skipping private key migration.")
```

File: lib/charms/observability_libs/v1/cert_handler.py [hunk 2]
```diff
@@ -349,21 +349,36 @@ def __init__(
...
                 return
 
             # we used to be on old juju: our secret stuff is in peer data
-            if peer_backend.retrieve():
+            if contents := peer_backend.retrieve():
+                logger.debug(
+                    "some data found in relation-backed vault. "
+                    "Migrating to secret-backed vault..."
```
Comment (sed-i) [inline lib/charms/observability_libs/v1/cert_handler.py:378]:
```suggestion
                    "Private key found in relation data. "
                    "Migrating private key to a juju secret."
```

File: lib/charms/observability_libs/v1/cert_handler.py [hunk 3]
```diff
@@ -349,21 +349,36 @@ def __init__(
...
     def _on_upgrade_charm(self, _):
+        has_privkey = self.vault.get_value("private-key")
+
         self._migrate_vault()
 
+        # If we already have a csr, but the pre-migration vault has no privkey stored,
+        # the csr must have been signed with a privkey that is now outdated and utterly lost.
+        # So we throw away the csr and generate a new one (and a new privkey along with it).
+        if not has_privkey and self._csr:
+            logger.debug("CSR and privkey out of sync after vault migration... renewing CSR.")
```
Comment (sed-i) [inline lib/charms/observability_libs/v1/cert_handler.py:360]:
```suggestion
            logger.debug("CSR and privkey out of sync after charm upgrade. Renewing CSR.")
```

**Review by sed-i — APPROVED** [review]

**Pull request merged**